aspnetcore/fundamentals/minimal-apis/bindStreamPipeReader/7.0-samples/PipeStreamToBackgroundQueue/Program.cs

```diff
@@ -0,0 +1,40 @@
+using System.Threading.Channels;
+using BackgroundQueueService;
+
+var builder = WebApplication.CreateBuilder(args);
+// The max memory we're willing to use for the upload endpoint on this instance
+var maxMemory = 500 * 1024 * 1024;
+
+// The max size of a single message, we're staying below the default LOH size of 85K
+var maxMessageSize = 80 * 1024;
+
+// The max size of the queue based on those restrictions
+var maxQueueSize = maxMemory / maxMessageSize;
+
+// Create a channel to send data to the background queue.
+builder.Services.AddSingleton<Channel<ReadOnlyMemory<byte>>>((_) => Channel.CreateBounded<ReadOnlyMemory<byte>>(maxQueueSize));
+
+// Create a background queue service.
+builder.Services.AddHostedService<BackgroundQueue>();
+var app = builder.Build();
+
+// curl --request POST 'http://localhost:5256/register' --header 'Content-Type: application/json' --data-raw '{ "Name":"Samson", "Age": 23, "Country":"Nigeria" }'
+app.MapPost("/register", async (Stream body, Channel<ReadOnlyMemory<byte>> queue) =>
+{
+    // Create a rewindable stream to be able to reuse the body stream.
+    var reusableStream = new MemoryStream();
+
+    // Copy the request body to the reusable stream.
+    await body.CopyToAsync(reusableStream);
+
+    // get buffer to avoid double allocation
+    reusableStream.TryGetBuffer(out var buffer);
+ 
+    // Send the buffer to the background queue.
+    await queue.Writer.WriteAsync(buffer);
+    
+    return Results.Accepted();
+});
```

Comment:
```suggestion
app.MapPost("/register", async (HttpRequest req, Stream body, Channel<ReadOnlyMemory<byte>> queue) =>
{
    if (req.ContentLength is not null && req.ContentLength > maxMessageSize)
    {
        // Message size exceeded
        return Results.BadRequest();
    }

    // We're not above the message size and we have a content length, or
    // we're a chunked request and we're going to read up to the maxMessageSize + 1. 
    // We add one to the message size so that we can detect when a chunked request body
    // is bigger than our configured max.
    var readSize = (int?)req.ContentLength ?? (maxMessageSize + 1);

    var buffer = new byte[readSize];

    // Read at least that many bytes from the body
    var read = await body.ReadAtLeastAsync(buffer, readSize, throwOnEndOfStream: false);

    // We read more than the max, so this is a bad request
    if (read > maxMessageSize)
    {
        // Message size exceeded
        return Results.BadRequest();
    }

    // Attempt to send the buffer to the background queue.
    if (queue.Writer.TryWrite(buffer.AsMemory(0..read)))
    {
        return Results.Accepted();
    }

    // We couldn't accept the message since we're overloaded
    return Results.StatusCode(StatusCodes.Status429TooManyRequests);
});
```

Untested, you'll need to run this for me, but I think it works. Review my codez please.

Reply:
ok. let me do that

Reply:
yh it works

Reply:
🥂

Reply:
please is this necessary ?
```
   // We read more than the max, so this is a bad request
    if (read > maxMessageSize)
    {
        // Message size exceeded
        return Results.BadRequest();
    }
```

doesn't this handle it
```
    if (req.ContentLength is not null && req.ContentLength > maxMessageSize)
    {
        // Message size exceeded
        return Results.BadRequest();
    }
```

Reply:
Also, I don't think the +1 is needed👀

Reply:
If the request is chunked and the content length is null, then no. I think ReadAtLeast can only read as much as the buffer that was passed in.

With your curl test, try adding `-H "Transfer-Encoding: chunked"` to try that code path.

Reply:
Okay thanks. I would try it out. Wasn't familiar with the chunked request 😅

Reply:
Ok, I just tried it now. And it worked when chunked (the content length was null). But my question is I'm not really clear on the relevance of the +1.

Reply:
> if (req.ContentLength is not null && req.ContentLength > maxMessageSize)
>     {
>         // Message size exceeded
>         return Results.BadRequest();
>     }

Also, should I remove this since, ReeadAtLeast can't read more than the buffer that was passed in

Reply:
Also, should I remove this since ReadAtLeast can't read more than the buffer that was passed in
```
    if (req.ContentLength is not null && req.ContentLength > maxMessageSize)
    {
        // Message size exceeded
        return Results.BadRequest();
    }
```

Reply:
You need `ReadAtLeast` because you need to make sure the the buffer is filled without having to manually write a stream reading loop. With `ReadAsync`, a stream today can return less than requested, we only want to do that if we're at the end of the stream. Here's the scenario you want to test, chunked encoding plus sending a payload > 80K. That's why the +1 is needed. If the message is > 80K then we want to throw and tell the user they're sending too much information.

Reply:
Ooh, that makes sense now. Thanks, sensei @davidfowl. Learnt something new